lib/Map/Vector/ProtomapsGeojsonSource.ts

```diff
@@ -0,0 +1,190 @@
+import Point from "@mapbox/point-geometry";
+import geojsonvt from "geojson-vt";
+import { makeObservable, observable, runInAction } from "mobx";
+import {
+  Bbox,
+  GeomType,
+  Feature as ProtomapsFeature,
+  TileSource,
+  Zxy
+} from "protomaps-leaflet";
+import {
+  FeatureCollectionWithCrs,
+  toFeatureCollection
+} from "../../ModelMixins/GeojsonMixin";
+import {
+  PROTOMAPS_DEFAULT_TILE_SIZE,
+  PROTOMAPS_TILE_BUFFER
+} from "../ImageryProvider/ProtomapsImageryProvider";
+
+type GeojsonVtFeature = {
+  id: any;
+  type: GeomType;
+  geometry: [number, number][][] | [number, number][];
+  tags: any;
+};
+
+type GeojsonVtTile = {
+  features: GeojsonVtFeature[];
+  numPoints: number;
+  numSimplified: number;
+  numFeatures: number;
+  source: any;
+  x: number;
+  y: number;
+  z: number;
+  transformed: boolean;
+  minX: number;
+  minY: number;
+  maxX: number;
+  maxY: number;
+};
+
+/** Extent (of coordinates) of tiles generated by geojson-vt */
+const geojsonvtExtent = 4096;
+
+/** Layer name to use with geojson-vt
+ *  This must be used in PaintRules/LabelRules (eg `dataLayer: "layer"`)
+ */
+export const GEOJSON_SOURCE_LAYER_NAME = "layer";
+
+export class ProtomapsGeojsonSource implements TileSource {
+  /** Data object from Options */
+  private readonly data: string | FeatureCollectionWithCrs;
+
+  /** Resolved geojsonObject (if applicable) */
+  @observable.ref
+  geojsonObject: FeatureCollectionWithCrs | undefined;
+
+  /** Geojson-vt tileIndex (if applicable) */
+  tileIndex: Promise<any> | undefined;
```

Comment:
Can the `tileIndex` get a more precise type than any, or does that information flow from untyped code?